### PR TITLE
Commonly used adaptation methods

### PR DESCRIPTION
**Pull request by LeahPrice**
Adding a class containing commonly used parameters and methods to adapt them (for static Bayesian models).

### CODE REVIEW

File: src/staticModelAdapt.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /// Chooses the next temperature such that a desired conditional ESS is maintained.
+    ///
+    /// \param logweight An armadillo vector containing the logarithm of the current particle weights.
+    /// \param loglike An armadillo vector containing the log likelihood of the current particle values.
+    /// \param desiredCESS The target conditional ESS for the next temperature (generally fixed).
+    void staticModelAdapt::ChooseTemp(const arma::vec & logweight, const arma::vec & loglike, double desiredCESS) {
+        double temp_curr = temp.back();
+        if (CESSdiff(logweight,loglike,1-temp_curr,desiredCESS)>=0){
```
Comment (adamjohansen) [inline src/staticModelAdapt.cpp]: Might it make sense for the comparison to be with -0.01 (or the tolerance) to be consistent with the inner workings of the bisection function?

File: src/staticModelAdapt.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        double f_a = CESSdiff(logweight,loglike,a-curr,desiredCESS);
+        double f_b = CESSdiff(logweight,loglike,b-curr,desiredCESS);
+        if (f_a*f_b>0){
+            Rcpp::stop("Bisection method to choose the next temperature failed");
+        } else{
+            double m, f_m, err;
+            m = (a+b)/2.0;
+            f_m = CESSdiff(logweight,loglike,m-curr,desiredCESS);
+            err = 10;
+            while (err > 0.01){
```
Comment (adamjohansen) [inline src/staticModelAdapt.cpp]: It might be good if it were possible to choose this tolerance without changing a hard-coded constant. In some applications the desired CESS might be larger than 0.99 and a tolerance of 0.01 would be problematic in those settings.

File: src/staticModelAdapt.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        cholCov = arma::chol(empCov);
+    }
+
+    /// Calculates the number of MCMC repeats based on the results of the most recent set of MCMC moves.
+    ///
+    /// \param acceptProb The proportion of accepted MCMC steps in the most recent iteration.
+    /// \param desiredAcceptProb The desired probability of a successful move for each particle.
+    /// \param initialN The initial number of MCMC repeats.
+    /// \param maxReps The maximum number of MCMC repeats.
+    int staticModelAdapt::calcMcmcRepeats(double acceptProb, double desiredAcceptProb, int initialN, int maxReps){
```
Comment (adamjohansen) [inline src/staticModelAdapt.cpp]: It's a bad idea to compare doubles with integers using ==, usually (the finite precision issue), might it make sense to replace these comparisons with 
abs(acceptProb - 1) <= 10**-9
or similar?

File: src/staticModelAdapt.cpp (hunk 2 above)
Comment (LeahPrice) [inline src/staticModelAdapt.cpp]: Thanks Adam for this and for all of your other good points. I should have picked up on at least some of these myself -- sorry about that. I've changed this and a couple of other minor things (like writing 1.0 instead of 1 if it's a double) and will do another commit when the check finishes.

**Comment by adamjohansen** [comment]
Great, thanks.
